cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -34,6 +34,139 @@ namespace cudf {
 namespace strings {
 namespace detail {
 
+// Strategy 1: String-parallel
+// This strategy assigns strings to warps so that each warp can cooperatively copy from the input
+// location of the string to the corresponding output location. Large datatype (uint4) is used for
+// stores. This strategy is best suited for large strings.
+
+// Helper function for loading 16B from a potentially unaligned memory location to registers.
+__forceinline__ __device__ uint4 load_uint4(const char* ptr)
+{
+  unsigned int* aligned_ptr = (unsigned int*)((size_t)ptr & ~(3));
+  uint4 regs                = {0, 0, 0, 0};
+
+  regs.x    = aligned_ptr[0];
+  regs.y    = aligned_ptr[1];
+  regs.z    = aligned_ptr[2];
+  regs.w    = aligned_ptr[3];
+  uint tail = aligned_ptr[4];
+
+  unsigned int shift = ((size_t)ptr & 3) * 8;
```

Comment:
Just use a modulo, it will get optimized to be the same as a bitwise operation and is more readable. 
```suggestion
  auto const shift = (static_cast<std::uintptr_t>(ptr) % 4) * 8;
```

Reply:
Just a heads-up, `static_cast` cannot convert pointer to integers. Will use `reinterpret_cast`.

Agree it's better to use modulo here.

But for
```
ptr & (~3)
```
do you think it's better to convert that to
```
ptr / 4 * 4
```

Reply:
Right, `reinterpret_cast` is necessary, my bad.

For `ptr & (~3)`, can't you just do `ptr/16`?

Reply:
I don't think so. `ptr & (~3)` is intended to set the least significant two bits to 0. For positive integers, that will be equivalent to the largest multiple of 4 smaller than `ptr`, hence my proposal of `ptr / 4 * 4`. Or `ptr - ptr % 4` would work as well.

Reply:
Sorry, misunderstood your original question.

This is how I'd write code like this. Store away the modulo result and reuse it:
```
  auto const offset = reinterpret_cast<uintptr_t>(ptr) % 4;
  auto const* aligned_ptr = reinterpret_cast<uint32_t*>(ptr - offset);
  uint4 regs                = {aligned_ptr[0], aligned_ptr[1], aligned_ptr[2], aligned_ptr[3]};
  auto const shift = offset * 8;
```